RecoEgamma/EgammaTools/src/HGCalShowerShapeHelper.cc

```diff
@@ -0,0 +1,310 @@
+  subDet_ = subDet;
+
+  setFilteredHitsAndFractions(hitsAndFracs);
+
+  setLayerWiseStuff();
+}
+
+void HGCalShowerShapeHelper::setPFRecHitPtrMap(const std::vector<reco::PFRecHit> &recHits) {
+  pfRecHitPtrMap_.clear();
+
+  for (auto &recHit : recHits) {
+    pfRecHitPtrMap_[recHit.detId()] = &recHit;
+  }
+}
+
+void HGCalShowerShapeHelper::setFilteredHitsAndFractions(const std::vector<std::pair<DetId, float> > &hitsAndFracs) {
+  hitsAndFracs_.clear();
+  hitEnergies_.clear();
+  hitEnergiesWithFracs_.clear();
+
+  for (const auto &hnf : hitsAndFracs) {
+    DetId hitId = hnf.first;
+    float hitEfrac = hnf.second;
+
+    int hitLayer = recHitTools_.getLayer(hitId);
+
+    if (hitLayer > nLayer_) {
+      continue;
+    }
+
+    if (hitId.det() != subDet_) {
+      continue;
+    }
+
+    if (pfRecHitPtrMap_.find(hitId.rawId()) == pfRecHitPtrMap_.end()) {
+      continue;
+    }
+
+    reco::PFRecHit recHit = *pfRecHitPtrMap_[hitId.rawId()];
+
+    if (recHit.energy() < minHitE_) {
+      continue;
+    }
+
+    if (recHit.pt2() < minHitET2_) {
+      continue;
+    }
+
+    // Fill the vectors
+    hitsAndFracs_.push_back(hnf);
+    hitEnergies_.push_back(recHit.energy());
+    hitEnergiesWithFracs_.push_back(recHit.energy() * hitEfrac);
+  }
+}
+
+void HGCalShowerShapeHelper::setLayerWiseStuff() {
+  layerEnergies_.clear();
+  layerEnergies_.resize(nLayer_);
+
+  layerCentroids_.clear();
+  layerCentroids_.resize(nLayer_);
+
+  centroid_.SetXYZ(0, 0, 0);
+
+  int iHit = -1;
+  double totalW = 0;
+
+  // Compute the centroid per layer
+  for (const auto &hnf : hitsAndFracs_) {
+    iHit++;
+
+    DetId hitId = hnf.first;
+
+    double weight = hitEnergies_[iHit];
+    totalW += weight;
+
+    const auto &hitPos = recHitTools_.getPosition(hitId);
+    ROOT::Math::XYZVector hitXYZ(hitPos.x(), hitPos.y(), hitPos.z());
+
+    centroid_ += weight * hitXYZ;
+
+    int hitLayer = recHitTools_.getLayer(hitId) - 1;
+
+    layerEnergies_[hitLayer] += weight;
+    layerCentroids_[hitLayer] += weight * hitXYZ;
+  }
+
+  int iLayer = -1;
+
+  for (auto &centroid : layerCentroids_) {
+    iLayer++;
+
+    if (layerEnergies_[iLayer]) {
+      centroid /= layerEnergies_[iLayer];
+    }
+  }
+
+  if (totalW) {
+    centroid_ /= totalW;
+  }
+}
+
+double HGCalShowerShapeHelper::getCellSize(DetId detId) {
+  double siThickness = recHitTools_.getSiThickness(detId);
+
+  // HD wafers
+  if (siThickness < 150) {
+    return 0.465;
+  }
+
+  // LD wafers
+  else {
+    return 0.698;
+  }
+}
+
+double HGCalShowerShapeHelper::getRvar(double cylinderR, double energyNorm, bool useFractions, bool useCellSize) {
+  if (hitsAndFracs_.empty()) {
+    return 0;
+  }
+
+  assert(energyNorm > 0);
+
+  double cylinderR2 = cylinderR * cylinderR;
+
+  double Rvar = 0;
+
+  std::vector<double>::iterator hitEnergyIter;
+
+  if (useFractions) {
+    hitEnergyIter = hitEnergiesWithFracs_.begin();
+  }
+
+  else {
+    hitEnergyIter = hitEnergies_.begin();
+  }
```

Comment:
would have replaced with auto hitEnergyIter  = useFractions ? hitEnergiesWithFracs_.begin() :  hitEnergies_.begin();

regardless screwing around with iterators like this will probably not pass review but we shall see..